### PR TITLE
add counters scenarios

### PR DESCRIPTION
**Pull request by aviv86**

### CODE REVIEW

File: Scenarios/Counters/Program.cs [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+    class Program
+    {
+        static void Main(string[] args)
+        {
+            RunCountersScenarios(args[0]);
+        }
+
+        static void RunCountersScenarios(string orchestratorUrl)
+        {
+            var halfHour = TimeSpan.FromMinutes(30);
```
Comment (aviviadi) [inline Scenarios/Counters/Program.cs]: Please remove the half hour loop. The orchestrator will execute the test in loop/parallel later on.

File: Scenarios/Counters/Properties/launchSettings.json [hunk 2]
```diff
@@ -0,0 +1,8 @@
+{
```
Comment (aviviadi) [inline Scenarios/Counters/Properties/launchSettings.json]: Please remove this from the PR (you may add launchSettings.json to .gitignore)

File: Scenarios/Counters/PutCommentsTest.cs [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+
+        private void PutBlogCommentDocs()
+        {
+            using (var bulk = DocumentStore.BulkInsert())
+            {
+                // bulk insert 1000 BlogComment docs
+
+                for (int i = 0; i < 1024; i++)
+                {
+                    var rnd = new Random();
```
Comment (aviviadi) [inline Scenarios/Counters/PutCommentsTest.cs]: use new https://github.com/aviviadi/TestingEnvironment/blob/master/TestingEnvironment.Client/BaseTest.cs#L17

File: Scenarios/Counters/PutCountersOnCommentsBasedOnTopic.cs [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+            {
+                var viewed = new List<string>();
+                var liked = new List<string>();
+                var disliked = new Dictionary<string, bool>();
+
+                using (var session = DocumentStore.OpenSession())
+                {
+                    // query random 128 BlogComment docs
+
+                    var rand = new Random();
```
Comment (aviviadi) [inline Scenarios/Counters/PutCountersOnCommentsBasedOnTopic.cs]: use new https://github.com/aviviadi/TestingEnvironment/blob/master/TestingEnvironment.Client/BaseTest.cs#L17

File: Scenarios/Counters/PutCountersOnCommentsRandomly.cs [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+            {
+                var viewed = new List<string>();
+                var liked = new List<string>();
+                var disliked = new List<string>();
+
+                using (var session = DocumentStore.OpenSession())
+                {
+                    // query random 128 BlogComment docs
+
+                    var rand = new Random();
```
Comment (aviviadi) [inline Scenarios/Counters/PutCountersOnCommentsRandomly.cs]: use new https://github.com/aviviadi/TestingEnvironment/blob/master/TestingEnvironment.Client/BaseTest.cs#L17

File: Scenarios/Counters/Program.cs (hunk 1 above)
Comment (aviv86) [reply Scenarios/Counters/Program.cs]: done

File: Scenarios/Counters/Properties/launchSettings.json (hunk 2 above)
Comment (aviv86) [reply Scenarios/Counters/Properties/launchSettings.json]: done

File: Scenarios/Counters/PutCommentsTest.cs (hunk 3 above)
Comment (aviv86) [reply Scenarios/Counters/PutCommentsTest.cs]: done

File: Scenarios/Counters/PutCountersOnCommentsBasedOnTopic.cs (hunk 4 above)
Comment (aviv86) [reply Scenarios/Counters/PutCountersOnCommentsBasedOnTopic.cs]: done

File: Scenarios/Counters/PutCountersOnCommentsRandomly.cs (hunk 5 above)
Comment (aviv86) [reply Scenarios/Counters/PutCountersOnCommentsRandomly.cs]: done

File: Scenarios/Counters/PatchCommentRatingsBasedOnCounters.cs [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+
+                try
+                {
+                    op.WaitForCompletion(TimeSpan.FromMinutes(1));
+                }
+                catch (TimeoutException)
+                {
+                    try
+                    {
+                        op.WaitForCompletion(TimeSpan.FromMinutes(1));
```
Comment (ayende) [inline Scenarios/Counters/PatchCommentRatingsBasedOnCounters.cs]: Why not just wait once for 2 minutes?

File: Scenarios/Counters/Program.cs [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+    class Program
+    {
+        static void Main(string[] args)
+        {
+            RunCountersScenarios(args[0]);
+        }
+
+        static void RunCountersScenarios(string orchestratorUrl)
+        {
+            using (var client = new PutCommentsTest(orchestratorUrl, "PutCommentsTest"))
```
Comment (ayende) [inline Scenarios/Counters/Program.cs]: Why do we need to repeat the class name each time?

File: Scenarios/Counters/Program.cs [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+            }
+
+            using (var client =
+                new PutCountersOnCommentsBasedOnTopic(orchestratorUrl, "PutCountersOnCommentsBasedOnTopic"))
+            {
+                client.Initialize();
+                client.RunTest();
+            }
+
+            Thread.Sleep(10000);
```
Comment (ayende) [inline Scenarios/Counters/Program.cs]: Why is this here?

File: Scenarios/Counters/PutCommentsTest.cs [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+                var numOfRetries = 3;
+                while (true)
+                {
+                    try
+                    {
+                        PutBlogCommentDocs();
+                    }
+                    catch (Exception e)
+                    {
+                        if (--numOfRetries > 0)
```
Comment (ayende) [inline Scenarios/Counters/PutCommentsTest.cs]: Let's report the errors as well

Comment (ayende) [reply Scenarios/Counters/PutCommentsTest.cs]: What kind of issues are we expecting here?

File: Scenarios/Counters/PutCountersOnCommentsRandomly.cs [hunk 10]
```diff
@@ -0,0 +1,134 @@
...
+                    var count = query.Count();
+                    var take = 128;
+
+                    var skip = count - take < 0
+                        ? 0
+                        : Random.Next(0, count - take);
+
+                    ReportInfo("Started querying random 128 docs");
+
+                    var streamQuery = session.Advanced.Stream(query.Skip(skip).Take(take));
```
Comment (ayende) [inline Scenarios/Counters/PutCountersOnCommentsRandomly.cs]: This is stressing the deep paging code, is that intentional?

File: Scenarios/Counters/QueryBlogCommentsAndIncludeCounters.cs [hunk 11]
```diff
@@ -0,0 +1,108 @@
...
+                    while (true)
+                    {
+                        try
+                        {
+                            result = query.ToList();
+                            break;
+                        }
+                        catch (Exception e)
+                        {
+                            if (--retries > 0)
```
Comment (ayende) [inline Scenarios/Counters/QueryBlogCommentsAndIncludeCounters.cs]: Report the exception here, and what kind of error are we expecting here?
